### DISCUSSION TITLE
tutorials/editor/external_editor

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#31] [User-contributed notes]
# tutorials/editor/external_editor

This page explains how to code using an external text editor. Godot can be used with an external text editor, such as Sublime Text or Visual Studio Code. Browse to the relevant editor settings: Edi...

https://docs.godotengine.org/en/latest/tutorials/editor/external_editor.html

**Comment by bleedweedsuz** [comment]
Godot can be used with an external text editor, such as Sublime Text or Visual Studio Code. Browse to the relevant editor settings: Editor > Editor Settings > Text Editor > External

this is change in MAC its in Godot > Editor Settings...

**Comment by p10tr3k** [comment]
# Integrating Godot with VSCode using URI scheme instead of CLI command (no lag)
In Godot, when you use VSCode as an external editor and rely on CLI commands with arguments to open files in VSCode on macOS, there's some lag until the file is opened. In the Dock, a second VSCode icon appears momentarily before the file opens, which takes a second or two. After some research, I found that a better approach to integrate with VSCode is to use the vscode://file/ URI scheme instead of CLI commands. This fixed my issue—there is no more lag when opening scripts from Godot in the VSCode editor, and the additional VSCode icon no longer appears in the Dock.

Here's how you can set it up:
## Create a Bash script:
```
touch ~/open_in_vscode_uri.sh 
chmod +x ~/open_in_vscode_uri.sh 
nano ~/open_in_vscode_uri.sh
```
## Edit the script:
```
#!/bin/bash

FILE_PATH="$1"
LINE="$2"
COLUMN="$3"

open "vscode://file/${FILE_PATH}:${LINE}:${COLUMN}"
```

##  Configure Godot:

* Open Godot.
* Go to Editor > Editor Settings.
* In the Text Editor > External section, set the Exec Path to the path of the script, e.g., ~/open_in_vscode_uri.sh.
* In the Exec Flags field, enter: "{file}" {line} {col}

[reference](https://www.reddit.com/r/godot/comments/1d73syh/integrating_godot_with_vscode_using_uri_scheme/)

**Reply by zastrowm** [reply]
This is so much faster.

Using a relative path to the bash script didn't work for me - godot wouldn't launch scripts in that case.  Changing it to an absolute path got it working.

**Reply by artemudovyk** [reply]
If you are on Linux, replace `open` with `xdg-open`:
```
xdg-open "vscode://file/${FILE_PATH}:${LINE}:${COLUMN}"
```

But for me the speed was the same, so I've used the method from the documentation instead.

**Reply by DanTulovsky** [reply]
If you want your entire project opened, claude came up with this monstrosity...  apparently the URI syntax doesn't support project and goto file at the same time...

```
#!/bin/bash

FILE_PATH="$1"
LINE="$2"
COLUMN="$3"

# Get the Godot project directory
PROJECT_DIR=$(dirname "$FILE_PATH")

# Traverse up to find project.godot
while [[ "$PROJECT_DIR" != "/" ]]; do
    if [[ -f "$PROJECT_DIR/project.godot" ]]; then
        break
    fi
    PROJECT_DIR=$(dirname "$PROJECT_DIR")
done

# Open folder first, then goto file
open "vscode://file/${PROJECT_DIR}/" && sleep 0.2 && open "vscode://file/${FILE_PATH}:${LINE}:${COLUMN}"
```

But it does work...

**Comment by lypoluz** [comment]
In case an emacs server is used, the command `emacsclient -f +{line}:{col} {file}` should be used. -f will allow emacs to reuse an existing frame instead of opening a new one.

**Comment by Qubus0** [comment]
if you are bothered by the popup telling you that a file was changed, open your Editor Settings and enable `text_editor/behavior/files/auto_reload_scripts_on_external_change`

**Reply by Calinou** [reply]
There's a PR enabling this by default:

- https://github.com/godotengine/godot/pull/97148

**Comment by flip-rossi** [comment]
Unfortunately, as Neovim's remote command line interface is not as complete as Vim's, some workaround is needed to  use the same terminal Neovim instance to keep opening files from Godot (and jumping to line and column).

I'm using this script:
```sh
#!/bin/sh

TERM_EXEC="alacritty -e"

if [ $# -ne 4 ]; then
    echo "USAGE: $0 <socket> <file> <line> <column>"
    exit 1
fi

SOCKET="$1"
FILE="$2"
LINE="$3"
COL="$4"

[ -S "$SOCKET" ] &&
    # Connect to running nvim server if socket exists
    nvim --server "$SOCKET" --remote-send ":n +call\ cursor($LINE,$COL) $FILE<CR>" || (
    # Create new server if socket doesn't exist
    tty -s && # Test if shell session is interactive, or a terminal should be opened
        nvim --listen "$SOCKET" "+call cursor($LINE,$COL)" "$FILE" || (
        nohup $TERM_EXEC nvim --listen "$SOCKET" "+call cursor($LINE,$COL)" "$FILE" > /dev/null 2>&1 & ) )
```

then, in Godot, I'm setting "Exec Path" to `sh` and "Exec Args" to `-c "~/Scripts/nvim-or-remote.sh ~/.cache/nvim/godot-$(basename {project}).socket {file} {line} {col}"`

**Reply by bdogabriel** [reply]
I modified your script to tcreate/attach a tmux session and restore the neovim session (with mini.sessions plugin):
```bash
#!/usr/bin/env bash

# config in editor:
# exec path: /bin/zsh
# exec flags: -c "source ~/.zshrc && ~/scripts/nvim-godot-remote.sh {project} ~/.cache/nvim/godot-$(basename {project}).socket {file} {line} {col}"

TERM_EXEC="wezterm start --"

if [ $# -ne 5 ]; then
	echo "USAGE: $0 <project dir> <socket> <file> <line> <column>"
	exit 1
fi

PROJECT_DIR="$1"
SOCKET="$2"
FILE="$3"
LINE="$4"
COL="$5"

PROJECT_NAME=$(basename "$PROJECT_DIR")

# Function to wait for socket to be available
wait_for_socket() {
	local socket="$1"
	local timeout=10
	local count=0

	while [ ! -S "$socket" ] && [ $count -lt $timeout ]; do
		sleep 0.5
		count=$((count + 1))
	done

	return $count
}

# Check if socket exists and try to connect first
if [ -S "$SOCKET" ]; then
	nvim --server "$SOCKET" --remote-send ":n +call\ cursor($LINE,$COL) $FILE<CR>"
else
	# Check if tmux session with project name exists
	if tmux has-session -t "$PROJECT_NAME" 2>/dev/null; then
		# Session exists - attach to it and open nvim
		if tty -s; then
			# We're in a terminal, attach directly
			tmux attach-session -t "$PROJECT_NAME" \; send-keys "cd '$PROJECT_DIR' && nvim --listen '$SOCKET'" C-m
		else
			# Not in terminal, open new terminal and attach
			nohup $TERM_EXEC tmux attach-session -t "$PROJECT_NAME" \; send-keys "cd '$PROJECT_DIR' && nvim --listen '$SOCKET'" C-m >/dev/null 2>&1 &
		fi
	else
		# Session doesn't exist - create new one
		if tty -s; then
			# We're in a terminal, create session directly
			tmux new-session -d -s "$PROJECT_NAME" -c "$PROJECT_DIR" "nvim --listen '$SOCKET'"
			tmux attach-session -t "$PROJECT_NAME"
		else
			# Not in terminal, open new terminal with new tmux session
			nohup $TERM_EXEC tmux new-session -s "$PROJECT_NAME" -c "$PROJECT_DIR" "nvim --listen '$SOCKET'" >/dev/null 2>&1 &
		fi
	fi
	# Wait for socket to be ready
	wait_for_socket "$SOCKET"
	# First restore session, then open the file
	nvim --server "$SOCKET" --remote-send ":lua require('mini.sessions').read('$PROJECT_NAME')<CR>"
	sleep 0.2 # Small delay to ensure session is loaded
	nvim --server "$SOCKET" --remote-send ":n +call\ cursor($LINE,$COL) $FILE<CR>"
fi
```

**Comment by NobleMode** [comment]
Sublime Text **Exec Flags** should be corrected from ``` {project} {file}:{line}:{column} ``` to ``` {project} {file}:{line}:{col} ``` according to the **Exec Flag Fields**. Using the old **Exec Flags** would cause Godot to point to Sublime Text incorrectly as ``` project example.gd:0:{column} ```

**Comment by SCAMX28** [comment]
With the **Kate** editor is better to use as exec flags: 
```
-s YourSession {file}:{line}:{col}
```
If you use the exec flags of this page the editor won´t update your code if you connect a signal through Godot.

**Reply by Calinou** [reply]
What should `-s YourSession` be set to? In particular, if you have multiple Godot editor instances open on different projects at once, I'd expect each instance to use its own session with a randomized ID.

This may require adding a new placeholder that Exec Flags can use, such as `{session_id}`.

**Comment by Calfur** [comment]
For Cursor on Mac (installed with Brew) these settings worked for me:

- External Editor: `Custom`
- Custom Exec Path: `/Applications/Cursor.app/Contents/MacOS/Cursor`
- Custom Exec Path Args: `{project} --goto {file}:{line}:{col}`

As I prefer to open the whole repo in the IDE and not just the Godot project which is in a subdir, I modified it then to this:

- Custom Exec Path Args: `{project}/../../ --goto {file}:{line}:{col}`

**Comment by Jeremi360** [comment]
What if I wants to use Rider but for GDScript?

**Comment by CodeTriangle** [comment]
For kakoune, you can achieve rich functionality with `kak +{line}:{col} {file}`. If you want every window to open in the same session, use `kak -c godot +{line}:{col} {file}` and start a headless kakoune session beforehand with `kak -d -s godot`.
